Complete Flutter VLC Player Rewrite - co-authroed with  @alr2413

mitchross opened the pull request:
https://github.com/solid-software/flutter_vlc_player/issues/158

@alr2413 and I have been hard at work past few months rewriting VLC plugin from scratch to support plugin v2. See https://medium.com/flutter/modern-flutter-plugin-development-4c3ee015cf5a

Here we have done a few things.

1. Leverages V2 Android Flutter Plugin design pattern.
2. All VLC features enabled.
3. Flutter Platform Interface to support any platform. The platform interface allows for platform specific implementation, ie web, desktop, linux, ios, android. Currently only iOS and Android are supported.
4. Flutter Pigeon for type safe method calls. No more string matching on method calls! (https://pub.dev/packages/pigeon)


Couple things.

We started a brand new repo for the work. It can be found here -> https://github.com/alr2413/flt_vlc_player/tree/master/flutter_vlc_player

This PR forks from Solid Software, incorporates @alr2413  code from their Repo and packages it into one singular PR commit.

Testing

1. If you switch to this branch/pull. You need to goto example folder inside flutter_vlc_folder , delete ios and android folder. Then Inside example run "flutter create . ", The git switch over is causing flutter a hard time in example to run it. Idk why.

solid-software reviewed (changes requested):
Great! Thank you!

Could you please restore funding.yaml and do some cleanup (remove deprecated/commented out files)?

Thanks for your awesome work!

solid-yuriiprykhodko commented:
Unable to run example app on iOS device:
```
$ pwd
/Users/yuriiprykhodko/IdeaProjects/flutter_vlc_player/flutter_vlc_player/example

$ flutter pub get
Running "flutter pub get" in example...                             0.5s

$ flutter run
Launching lib/main.dart on iPhone in debug mode...
Expected ios/Runner.xcodeproj/project.pbxproj but this file is missing.
No application found for TargetPlatform.ios.
Is your project missing an ios/Runner/Info.plist?
Consider running "flutter create ." to create one.
```

mitchross commented:
> Unable to run example app on iOS device:
> 
> ```
> $ pwd
> /Users/yuriiprykhodko/IdeaProjects/flutter_vlc_player/flutter_vlc_player/example
> 
> $ flutter pub get
> Running "flutter pub get" in example...                             0.5s
> 
> $ flutter run
> Launching lib/main.dart on iPhone in debug mode...
> Expected ios/Runner.xcodeproj/project.pbxproj but this file is missing.
> No application found for TargetPlatform.ios.
> Is your project missing an ios/Runner/Info.plist?
> Consider running "flutter create ." to create one.
> ```

Try this.... 

I cannot figure out why its doing this. The gitignore may be wrong... 

```
(base) mitchross@MacBook-Pro example % ls
README.md	build		flutter_02.log	pubspec.lock
android		example.iml	ios		pubspec.yaml
assets		flutter_01.log	lib		test
(base) mitchross@MacBook-Pro example % pwd
/Users/mitchross/Documents/Programming/flutter/flutter_vlc_player/flutter_vlc_player/example
(base) mitchross@MacBook-Pro example % rm -rf ios
(base) mitchross@MacBook-Pro example % flutter create .
Recreating project ....
```

Then in your git tool of choice discard all of the changes besides the xcworkspace changes. If you keep the files after the create it will fail.

<img width="571" alt="Screen Shot 2021-01-11 at 10 32 59 AM" src="https://user-images.githubusercontent.com/6330506/104202363-da09c400-53f8-11eb-91f8-adfe11ce490d.png">

Discard these 
<img width="1081" alt="Screen Shot 2021-01-11 at 10 33 58 AM" src="https://user-images.githubusercontent.com/6330506/104202405-e857e000-53f8-11eb-8704-72fee127717e.png">
<img width="1080" alt="Screen Shot 2021-01-11 at 10 33 53 AM" src="https://user-images.githubusercontent.com/6330506/104202406-e8f07680-53f8-11eb-9e30-55ff7f1126c9.png">
<img width="571" alt="Screen Shot 2021-01-11 at 10 33 03 AM" src="https://user-images.githubusercontent.com/6330506/104202409-e9890d00-53f8-11eb-9d89-b0fd6be96128.png">

solid-yuriiprykhodko commented:
Yeah, `git clean` and `rm -rf ios` did the trick, thanks!

mitchross commented:
> Yeah, `git clean` and `rm -rf ios` did the trick, thanks!

just pay attention to git changes. In the plist , you need to make sure the (red) in my screen shot is preserved and the extra files can be discarded... This will be a pain for users once they upgrade to 5.0, I just dont know the solve yet... maybe gitignore?

solid-yuriiprykhodko commented on flutter_vlc_player/lib/vlc_player_flutter.dart:
```diff
@@ -0,0 +1,37 @@
+        VlcStreamOutputOptions,
+        VlcVideoOptions;
+
+part 'src/enums/playing_state.dart';
+
+part 'src/vlc_player_controller.dart';
+
+part 'src/flutter_vlc_player.dart';
+
+part 'src/vlc_app_life_cycle_observer.dart';
+
+part 'src/vlc_player_value.dart';
```
Have you considered avoiding `part` and `part of` directives? They tend to make code navigation a bit more messy than with only `import`'s. 

Also usage of `part` and `part of` is discouraged by the Dart team.
https://dart.dev/guides/libraries/create-library-packages#organizing-a-library-package

mitchross replied:
Yea this shouldnt be a big deal to resolve, part of why we need code reviews! We forget about minute things when deep in the code.

alr2413 replied:
fixed

solid-yuriiprykhodko commented on flutter_vlc_player/android/build.gradle:
```diff
@@ -31,12 +31,17 @@ android {
+    implementation 'androidx.appcompat:appcompat:1.2.0'
     implementation 'androidx.legacy:legacy-support-v4:1.0.0'
     implementation 'androidx.annotation:annotation:1.1.0'
+    implementation 'org.jetbrains:annotations-java5:15.0'
```
Had to remove this line to fix build for Android. Otherwise I was getting the following error:
```
$ flutter run
Launching lib/main.dart on MI MAX 2 in debug mode...
Note: /Users/yuriiprykhodko/IdeaProjects/flutter_vlc_player/flutter_vlc_player/android/src/main/java/software/solid/fluttervlcplayer/FlutterVlcPlayer.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.                    
                                                                        
FAILURE: Build failed with an exception.                                
                                                                        
* What went wrong:                                                      
Execution failed for task ':app:checkDebugDuplicateClasses'.            
> 1 exception was raised by workers:                                    
  java.lang.RuntimeException: Duplicate class org.intellij.lang.annotations.Flow found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.intellij.lang.annotations.Identifier found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.intellij.lang.annotations.JdkConstants found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.intellij.lang.annotations.JdkConstants$AdjustableOrientation found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.intellij.lang.annotations.JdkConstants$BoxLayoutAxis found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.intellij.lang.annotations.JdkConstants$CalendarMonth found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.intellij.lang.annotations.JdkConstants$CursorType found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.intellij.lang.annotations.JdkConstants$FlowLayoutAlignment found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.intellij.lang.annotations.JdkConstants$FontStyle found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.intellij.lang.annotations.JdkConstants$HorizontalAlignment found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.intellij.lang.annotations.JdkConstants$InputEventMask found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.intellij.lang.annotations.JdkConstants$ListSelectionMode found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.intellij.lang.annotations.JdkConstants$PatternFlags found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.intellij.lang.annotations.JdkConstants$TabLayoutPolicy found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.intellij.lang.annotations.JdkConstants$TabPlacement found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.intellij.lang.annotations.JdkConstants$TitledBorderJustification found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.intellij.lang.annotations.JdkConstants$TitledBorderTitlePosition found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.intellij.lang.annotations.JdkConstants$TreeSelectionMode found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.intellij.lang.annotations.Language found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.intellij.lang.annotations.MagicConstant found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.intellij.lang.annotations.Pattern found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.intellij.lang.annotations.PrintFormat found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.intellij.lang.annotations.PrintFormatPattern found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.intellij.lang.annotations.RegExp found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.intellij.lang.annotations.Subst found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.jetbrains.annotations.Contract found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.jetbrains.annotations.Nls found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.jetbrains.annotations.NonNls found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.jetbrains.annotations.NotNull found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.jetbrains.annotations.Nullable found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.jetbrains.annotations.PropertyKey found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
  Duplicate class org.jetbrains.annotations.TestOnly found in modules annotations-13.0.jar (org.jetbrains:annotations:13.0) and annotations-java5-15.0.jar (org.jetbrains:annotations-java5:15.0)
                                                                        
  Go to the documentation to learn how to <a href="d.android.com/r/tools/classpath-sync-errors">Fix dependency resolution errors</a>.
                                                                        
                                                                        
* Try:                                                                  
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
                                                                        
* Get more help at https://help.gradle.org                              
                                                                        
BUILD FAILED in 47s                                                     
Running Gradle task 'assembleDebug'...                                  
Running Gradle task 'assembleDebug'... Done                        48.1s
Exception: Gradle task assembleDebug failed with exit code 1

```

solid-yuriiprykhodko commented on flutter_vlc_player/lib/src/vlc_player_controller.dart:
```diff
@@ -0,0 +1,883 @@
+    }
+    if (position > value.duration) {
+      position = value.duration;
+    } else if (position < const Duration()) {
+      position = const Duration();
+    }
```
Nit: we could be a bit more explicit here using `Duration.zero`
```suggestion
    } else if (position < Duration.zero) {
      position = Duration.zero;
    }
```

alr2413 replied:
fixed.

solid-yuriiprykhodko commented on flutter_vlc_player/ios/Classes/SwiftFlutterVlcPlayerPlugin.swift:
```diff
@@ -0,0 +1,1228 @@
+    
+    public func setLooping(_ input: LoopingMessage, error: AutoreleasingUnsafeMutablePointer<FlutterError?>) {
+        
+        var player: VLCViewController? = nil
+        if let textureId = input.textureId {
+            player = players[Int(truncating: textureId)]
+        }
```
This piece repeats a lot in methods here. Perhaps worth extracting it into a getter?

mitchross replied:
I tried this tonight, the issue is that the "input" variable has different types based on each method. Ie "_ input: TextureMessage",  "input: PositionMessage", "input: LoopingMessage" 

Im all for D.R.Y but not entirely sure how to manage here without using some sort of overkill generics. Thoughts?

solid-yuriiprykhodko replied:
We're grabbing the input by the textureId field everywhere, right? 
```swift
func getPlayer(textureId: Float?) -> VLCViewController? {
  guard textureId != nil else { return nil }
  return players[Int(truncating: textureId)]
}
```

mitchross replied:
Swift is not my native language :) I pushed up a change to review.

solid-yuriiprykhodko commented on flutter_vlc_player/example/lib/single_tab.dart:
```diff
@@ -0,0 +1,198 @@
+                    else
+                      Scaffold.of(context).showSnackBar(
+                        SnackBar(
+                          content: Text("File not exists!"),
```
Nit: "File does not exist!"

Looks like the option to play the video from file in the example app is not portable. 
I'm not sure we can ensure that a video is present on the file system when installing the app, but we can do that upon launch.

There are a couple of options here:
1. Include a small video as an asset and extract the video to filesystem upon launch.
2. Use file_picker or image_picker, so that the user could supply a video themselves.

I'm personally in favor of option 1. It will introduce a small delay, since the copying process isn't instant, but it still feels like a 'batteries included' example, compared to option 2, where the user will have to ensure they have a video somewhere on the filesystem.

mitchross replied:
What's the pseudo code for option 1 ?

solid-yuriiprykhodko replied:
Something like this:
```dart
void initState() {
  _fsVideoFuture = _loadVideoToFs();
}

Future<File> _loadVideoToFs() async {
  final videoData = await rootBundle.load('fs-video');
  final videoBytes = Uint8List.view(videoData.buffer);
  final dir = Directory.systemTemp;
  final videoFile = File(join(dir, 'fs-video.mp4'))
    ..createSync()
    ..writeAsBytesSync(videoBytes);
  return videoFile;
}
```

We can use the Future show a progress indicator while the video is loading.

mitchross replied:
we will try it!

mitchross replied:
Have this mostly working, discussing with @alr2413 on the side before I push up. Will finish up tomorrow.

alr2413 commented on flutter_vlc_player/example/lib/main.dart:
```diff
@@ -0,0 +1,47 @@
+
+void main() {
+  runApp(
+    MaterialApp(
+      home: App(),
+    ),
+  );
+}
+
+class App extends StatefulWidget {
+  @override
+  _AppState createState() => _AppState();
+}
+
+class _AppState extends State<App> {
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp(
```
nice catch.
fix that.

alr2413 commented on flutter_vlc_player_platform_interface/lib/src/platform_interface/vlc_player_platform_interface.dart:
```diff
@@ -0,0 +1,337 @@
+  /// This method is called when the plugin is first initialized
+  /// and on every full restart.
+  Future<void> init() {
+    throw UnimplementedError('init() has not been implemented.');
```
Ok. 
fix it.

solid-yuriiprykhodko commented on flutter_vlc_player/lib/src/vlc_player_value.dart:
```diff
@@ -134,6 +134,10 @@ class VlcPlayerValue {
     return aspectRatio;
   }
 
+   void setInitialized(bool initialized) {
+    _initialized = initialized;
+  }
+
```
Not sure we want to expose this as a public method, since this class is exported to the plugin user. 

Making it a part of the `copyWith` would make the code a bit more consistent - it is used as an internal method for updating `VlcPlayerValue` fields. It's still part of the public interface, but using `copyWith` at least won't have any unexpected results in the code, since the user can't use the copied value anywhere in the plugin. 
At the same time updating the field on the player value can be read from elsewhere in the program, leading to undefined behavior.

alr2413 replied:
i agree with you.

alr2413 replied:
fixed

mitchross replied:
Agree. these were side effects of removing part files. I was trying something to get the code to compile and pushed up for @alr2413. We probably should have worked in a side branch, when I pushed, I realized there was a risk that you would see it before we had a chance to talk about it.

solid-yuriiprykhodko commented on flutter_vlc_player/lib/src/vlc_player_controller.dart:
```diff
@@ -29,6 +31,10 @@ typedef RendererCallback = void Function(VlcRendererEventType, String, String);
 ///
 /// After [dispose] all further calls are ignored.
 class VlcPlayerController extends ValueNotifier<VlcPlayerValue> {
+  VlcPlayerPlatform getVlcPlayerPlatform() {
+    return _vlcPlayerPlatform;
+  }
```
We don't need to expose this as a public method.

We can put the platfrom global variable (`_vlcPlayerPlatform`) into a separate file and make it public, and then just don't export it in `lib/vlc_flutter_player.dart`.

alr2413 replied:
fixed

solid-yuriiprykhodko commented on flutter_vlc_player/lib/src/vlc_player_controller.dart:
```diff
@@ -114,11 +120,19 @@ class VlcPlayerController extends ValueNotifier<VlcPlayerValue> {
+  void setViewId(viewId) {
+    _viewId = viewId;
+  }
+
```
The two added methods break encapsulation - severely. 

They are only used in the `onPlatformViewCreated` listener in the video player Widget.
At the very least we could make that listener a part of this class. This way it would be a bit more clear that it's a listener, and not a user-facing method.

On a second thought, this `onPlatformViewCreated` callback can be used by someone who would want to make a custom widget with the controller. We could leave it public, and add a doc comment explaining how it can be used.

alr2413 replied:
fixed

solid-yuriiprykhodko commented on flutter_vlc_player/android/src/main/java/software/solid/fluttervlcplayer/FlutterVlcPlayer.java:
```diff
@@ -0,0 +1,612 @@
+                media = new Media(libVLC, context.getAssets().openFd(url));
+            else
+                media = new Media(libVLC, Uri.parse(url));
+            if (hwAcc != HwAcc.AUTOMATIC.getNumericType()) {
+                if (hwAcc == HwAcc.DISABLED.getNumericType()) {
+                    media.setHWDecoderEnabled(false, false);
+                } else if (hwAcc == HwAcc.FULL.getNumericType() || hwAcc == HwAcc.DECODING.getNumericType()) {
+                    media.setHWDecoderEnabled(true, true);
+                    if (hwAcc == HwAcc.DECODING.getNumericType()) {
+                        media.addOption(":no-mediacodec-dr");
+                        media.addOption(":no-omxil-dr");
+                    }
+                }
+            }
```
Nitpick: seems a bit convoluted. Could we turn it into a `switch`, perhaps?
```suggestion
            final HwAcc hwAcc = HwAcc.fromValue(hwAccValue);
            switch (hwAcc) {
                case DISABLED:
                    media.setHWDecoderEnabled(false, false);
                    break;
                case DECODING:
                case FULL:
                    media.setHWDecoderEnabled(true, true);
                    break;
            }
            if (hwAcc == HwAcc.DECODING) {
                media.addOption(":no-mediacodec-dr");
                media.addOption(":no-omxil-dr");
            }
```

solid-software commented on flutter_vlc_player/CHANGELOG.md:
```diff
@@ -1,11 +1,3 @@
-## 4.0.3
-Update VLCKit for iOS and iOS. Cleanup example Pod file. Clean up example gradle. Credits to Mitch Ross (https://github.com/mitchross).
-Removed dispose calls on VlcPlayerController from VlcPlayer.
-Fix argument-less functions throwing FlutterMethodNotImplemented.
-
-## 4.0.2
-* Update Cocoapods version for VLCkit on iOS. This fixes issues with iOS 12 and Simulators. Credits to Mitch Ross (https://github.com/mitchross).
-
```
Could you please restore those versions for the reference?

mitchross replied:
yea was lost in the git switch over. Ill add em back no problem.

mitchross replied:
fixed

mitchross commented:
> Great! Thank you!
> 
> Could you please restore funding.yaml and do some cleanup (remove deprecated/commented out files)? Also we seem to have lost some `changelog` versions and maybe some other changes we added while you were working on this?
> 
> Thanks for your awesome work!

fixed

solid-software commented on flutter_vlc_player/CHANGELOG.md:
```diff
@@ -1,3 +1,6 @@
+## 5.0.0
+Entire rewrite of Flutter VLC Player. Updates to Android v2 plugin. Adds Platform interface. Adds Pigeon for type safe method calls. Credits to credits to Alireza Setayesh (https://github.com/alr2413) and Mitch Ross (https://github.com/mitchross)
```
```suggestion
Entire rewrite of Flutter VLC Player. Updates to Android v2 plugin. Adds Platform interface. Adds Pigeon for type safe method calls. Credits to Alireza Setayesh (https://github.com/alr2413) and Mitch Ross (https://github.com/mitchross)
```

mitchross commented:
@solid-yuriiprykhodko 

Hoping you can help me identify a solution to the overlay bug seen on multi tab and on the single tab when switching over the url on iOS...
<img width="418" alt="Screen Shot 2021-01-13 at 9 26 50 PM" src="https://user-images.githubusercontent.com/6330506/104623693-9d400600-5660-11eb-9329-8c8503ba4774.png">


When you switch the stream, or load up a new VLCplayer and start playing again (ie hit change stream), the ios delegate never changes, ie mediaPlayerStateChanged never hits the .played switch case.

<img width="303" alt="Screen Shot 2021-01-13 at 9 27 04 PM" src="https://user-images.githubusercontent.com/6330506/104623764-b052d600-5660-11eb-9b4d-1c7ffcd49d75.png">

the first video works and we can see this sequence

<img width="304" alt="Screen Shot 2021-01-13 at 9 26 10 PM" src="https://user-images.githubusercontent.com/6330506/104623836-c496d300-5660-11eb-8507-b5a4659dd1f7.png">


Looks like there is a two year old bug and several days ago someone else reported this issue -> https://code.videolan.org/videolan/VLCKit/-/issues/144 .  

> " Malek @Red-Mak · 3 days ago
> This bug still not fixed, tested on MobileVLCKit / TVVLCKit 3.3.15 , is this in the roadmap ? for now im using mediaPlayerTimeChanged to hide loaders and do other stuff. Thanks.


The overlay code depends on this state to manage the visibility boolean. 

Im hoping you have ideas I can try to solve this. I am a bit stuck here...

mitchross commented:
@solid-yuriiprykhodko Also made an attempt at the load from file -> https://github.com/mitchross/flutter_vlc_player/pull/16 any suggestions?

mitchross commented:
> @solid-yuriiprykhodko Also made an attempt at the load from file -> [mitchross#16](https://github.com/mitchross/flutter_vlc_player/pull/16) any suggestions?

resolved

mitchross commented:
> @solid-yuriiprykhodko
> 
> Hoping you can help me identify a solution to the overlay bug seen on multi tab and on the single tab when switching over the url on iOS...
> <img alt="Screen Shot 2021-01-13 at 9 26 50 PM" width="418" src="https://user-images.githubusercontent.com/6330506/104623693-9d400600-5660-11eb-9329-8c8503ba4774.png">
> 
> When you switch the stream, or load up a new VLCplayer and start playing again (ie hit change stream), the ios delegate never changes, ie mediaPlayerStateChanged never hits the .played switch case.
> 
> <img alt="Screen Shot 2021-01-13 at 9 27 04 PM" width="303" src="https://user-images.githubusercontent.com/6330506/104623764-b052d600-5660-11eb-9b4d-1c7ffcd49d75.png">
> 
> the first video works and we can see this sequence
> 
> <img alt="Screen Shot 2021-01-13 at 9 26 10 PM" width="304" src="https://user-images.githubusercontent.com/6330506/104623836-c496d300-5660-11eb-8507-b5a4659dd1f7.png">
> 
> Looks like there is a two year old bug and several days ago someone else reported this issue -> https://code.videolan.org/videolan/VLCKit/-/issues/144 .
> 
> > " Malek @Red-Mak · 3 days ago
> > This bug still not fixed, tested on MobileVLCKit / TVVLCKit 3.3.15 , is this in the roadmap ? for now im using mediaPlayerTimeChanged to hide loaders and do other stuff. Thanks.
> 
> The overlay code depends on this state to manage the visibility boolean.
> 
> Im hoping you have ideas I can try to solve this. I am a bit stuck here...

@solid-pavloprykhodko , looks like @alr2413 has a solution. We are testing...

solid-software commented:
It looks like it's in a working state, what do you think if we merge this and then solve the issue in the another PR?

mitchross commented:
> It looks like it's in a working state, what do you think if we merge this and then solve the issue in the another PR?

@solid-software , the readme still needs to be updated. The ios example is a bit buggy, we are close to fixing. Give us a few hours, will get to mergeable state shortly. cc @alr2413

mitchross commented:
@solid-software @solid-pavloprykhodko  its ready for merge now. cc @alr2413

mitchross commented:
@solid-software .. cc @alr2413 Its come to our attention that this change cannot be published right away without some changes. If you try to manually point a new flutter project at this branch and do a flutter pub get, it will error out saying no pubspec was found. This is because the root folder name and the library folder name are the same.

There is a fix, we can take the contents of 'flutter_vlc_player' and move them to the root of the project. This will mostly work however there is one issue. You cant publish a package with path dependencies. Ie
```
 
   flutter_vlc_player_platform_interface:
          path: flutter_vlc_platform_interface
```
  
Instead we need

`  flutter_vlc_player_platform_interface: ^1.0.0`

However, we cant test this locally right now because flutter_vlc_player_platform_interface is not published yet. 

See -> https://dart.dev/tools/pub/dependencies

> Path dependencies are useful for local development, but do not work when sharing code with the outside world—not everyone can get to your file system. Because of this, you cannot upload a package to the pub.dev site if it has any path dependencies in its pubspec.

Refs - 
https://github.com/flutter/plugins/blob/master/packages/video_player/video_player/pubspec.yaml
https://github.com/flutter/plugins/blob/master/packages/video_player/video_player_platform_interface/pubspec.yaml

solid-yuriiprykhodko commented on flutter_vlc_player/README.md:
```diff
@@ -0,0 +1,115 @@
+
+### Version 5.0 Upgrade For Existing Apps
+To upgrade to version 5.0 first you need to migrate the existing project to swift.
+
+    First run 'git clean'
+    Delete existing ios folder from root of flutter project.
+    Run this command flutter create -i swift .
+    Be sure to follow instructions above after for plist and pod file changes.
+
```
Let's split out the commands as codeblocks - they're easier to read and copy that way. Example below:


1. Clean the repo:
```
git clean
```
2. Delete existing `ios` folder from root of flutter project. If you have some custom changes made to the iOS app - rename it or copy somewhere outside the project.
3. Re-create the iOS app:
```
flutter create -i swift .
```

Make sure to update the project according to warnings shown by the flutter tools. (Update Info.plist, Podfile).
If you have some changes made to the iOS app, recreate the app using above method and copy in the changed files.

mitchross replied:
fixed

solid-yuriiprykhodko commented on flutter_vlc_player/README.md:
````diff
@@ -0,0 +1,115 @@
+
+
+
+## API
+```dart
+TBD
+```
````
I'm not sure we need an API description in the README. Let's remove this section.

mitchross replied:
fixed

solid-yuriiprykhodko commented on flutter_vlc_player/README.md:
````diff
@@ -0,0 +1,115 @@
+To start using the plugin, copy this code or follow the example project in 'flutter_vlc_player/example'
+
+```dart
+TBD
````
Could you please add a minimal example here?

mitchross replied:
Added!

solid-yuriiprykhodko commented:
The player UI becomes unresponsive when it finishes playing. The Dart code seems to come through OK, `controller.play` gets dispatched. The problem is probably in the platform implementations. The bug shows on both Android and iOS.

mitchross commented:
> The player UI becomes unresponsive when it finishes playing. The Dart code seems to come through OK, `controller.play` gets dispatched. The problem is probably in the platform implementations. The bug shows on both Android and iOS.

Yes @alr2413 and I have been discussing...

mitchross commented:
> The player UI becomes unresponsive when it finishes playing. The Dart code seems to come through OK, `controller.play` gets dispatched. The problem is probably in the platform implementations. The bug shows on both Android and iOS.

I have fixed this and pushed. We can discuss if this is not best approach, but is working on both platforms now... cc @alr2413

mitchross commented:
> > The player UI becomes unresponsive when it finishes playing. The Dart code seems to come through OK, `controller.play` gets dispatched. The problem is probably in the platform implementations. The bug shows on both Android and iOS.
> 
> I have fixed this and pushed. We can discuss if this is not best approach, but is working on both platforms now :) ... cc @alr2413

Update 1pm 1/19/2021. Fully resolved.

yuopri reviewed (approved):
LGTM! 
Once again, great work!

solid-yuriiprykhodko reviewed (approved):
Sorry, wrong account. LGTM!

alr2413 commented:
Ping @solid-software,

The PR is already merged, but it's not released & not published in "pub.dev" repo.
Is there any issue?

solid-software commented:
Pong @alr2413 :)

No issues, we would probably want to merge other PRs as well before releasing the plugin:

https://github.com/solid-software/flutter_vlc_player/pull/162
https://github.com/solid-software/flutter_vlc_player/pull/161

Thanks again for your contribution!
